---
title: "**Autozygosity and duration of gestation.**"
author:
        - "Pol Sole-Navais"
date: "`r format(Sys.time(), '%d %B, %Y')`"
    
output:
    html_document:
        highlight: tango
        number_sections: no
        theme: default
        toc: yes
        toc_depth: 3
        toc_float:
            collapsed: no
            smooth_scroll: yes
---


```{r dependencies, include=FALSE}
library("ggplot2")
library("dplyr")
library("survival")
library("knitr")
library("gridExtra")
library("tidyr")
library("kableExtra")
library("survminer")
library("data.table")
options(warn=-1)
library(broom)
options(width = 1800)

```


```{r read_files, include= F}
input= snakemake@input

if (grepl('harvest', input[1])){

moms= unlist(input[grep('runs_mfr_maternal.txt', input)])
dads= unlist(input[grep('runs_mfr_paternal.txt', input)])
fets= unlist(input[grep('runs_mfr_fetal.txt', input)])
moms= read.table(moms, h=T, sep= '\t')
dads= read.table(dads, h=T, sep= '\t')
fets= read.table(fets, h=T, sep= '\t')
q1= unlist(input[grep('q1_pdb1724_v9.csv', input)])
q1= read.table(q1, h=T, sep= ';')

fmoms= unlist(input[grep('/results/het/maternal', input)])
fmoms= fread(fmoms)

fdads= unlist(input[grep('/results/het/paternal', input)])
fdads= fread(fdads)

ffets= unlist(input[grep('/results/het/fetal', input)])
ffets= fread(ffets)

moms= full_join(moms, fmoms, by= 'IID')
dads= full_join(dads, fdads, by= 'IID')
fets= full_join(fets, ffets, by= 'IID')


filter_df= function(df){
df= mutate(df, spont= as.numeric(FSTART==1 & (is.na(KSNITT) | KSNITT>1) &
                (is.na(KSNITT_PLANLAGT) | KSNITT_PLANLAGT==1) &
                INDUKSJON_PROSTAGLANDIN==0 &
                INDUKSJON_ANNET==0 &
                INDUKSJON_OXYTOCIN==0 &
                INDUKSJON_AMNIOTOMI==0),
                PARITY0= as.numeric(PARITET_5==0))

return(df)
}

q1$AA1124= as.integer(q1$AA1124)
q1$AA1125= as.integer(q1$AA1125)
q1$AA1128= as.integer(q1$AA1128)
q1$AA1129= as.integer(q1$AA1129)

q1$edu[q1$AA1124 <3 | q1$AA1125<3]= 1
q1$edu[q1$AA1124 ==3 | q1$AA1125==3]= 2
q1$edu[q1$AA1124 ==4 | q1$AA1125==4]= 2
q1$edu[q1$AA1124 ==5 | q1$AA1125==5]= 3
q1$edu[q1$AA1124 ==6 | q1$AA1125==6]= 4
q1$edu[(is.na(q1$edu) & q1$AA1128==6) | (is.na(q1$edu) & q1$AA1129==1)]= 5

q1$AA1315= as.integer(q1$AA1315)
q1$AA1316= as.integer(q1$AA1316)

q1$AA1315[q1$AA1315==1 | q1$AA1315==2| q1$AA1315==3| q1$AA1315==4]= 0
q1$AA1315[q1$AA1315==5 | q1$AA1315==6 | q1$AA1315==7]= 1
q1$AA1316[q1$AA1316==1 | q1$AA1316==2| q1$AA1316==3| q1$AA1316==4]= 0
q1$AA1316[q1$AA1316==5 | q1$AA1316==6 | q1$AA1316==7]= 1

q1$income[q1$AA1315==0 & q1$AA1316==0]= 0
q1$income[q1$AA1315==0 & is.na(q1$AA1316)]= 0
q1$income[is.na(q1$AA1315) & q1$AA1316==0]= 0
q1$income[q1$AA1315==1 | q1$AA1316==1]= 1
q1$income[q1$AA1315==1 & is.na(q1$AA1316)]= 1
q1$income[is.na(q1$AA1315) & q1$AA1316==1]= 1
q1$income[q1$AA1315==1 & q1$AA1316==1]= 2

q1$height= ifelse(q1$AA87>= 140 | q1$AA87<= 210, q1$AA87, NA)
q1= q1 %>% select(PREG_ID_1724, edu, income, height)
q1$edu= as.factor(q1$edu)

m= unlist(input[grep('harvest_maternal.hom', input)])
m= fread(m)

ms= m %>% group_by(CHR) %>% summarize(n= n(), KB= median(KB, na.rm=T))

d= unlist(input[grep('harvest_paternal.hom', input)])
d= fread(d)

ds= d %>% group_by(CHR) %>% summarize(n= n(), KB= median(KB, na.rm=T))

fe= unlist(input[grep('harvest_fetal.hom', input)])
fe= fread(fe)

fes= fe %>% group_by(CHR) %>% summarize(n= n(), KB= median(KB, na.rm=T))

moms= filter_df(moms)
dads= filter_df(dads)
fets= filter_df(fets)

moms_c= inner_join(moms, q1, by= 'PREG_ID_1724')
dads_c= inner_join(dads, q1, by= 'PREG_ID_1724')
fets_c= inner_join(fets, q1, by= 'PREG_ID_1724')

freqm= unlist(input[grep('runs/frequency/ROH_frequency_maternal', input)])
freqm= read.table(freqm, h=F, sep= '\t')
names(freqm)= c('chr','pos','freqm')

freqd= unlist(input[grep('runs/frequency/ROH_frequency_paternal', input)])
freqd= read.table(freqd, h=F, sep= '\t')
names(freqd)= c('chr','pos','freqd')

freqf= unlist(input[grep('runs/frequency/ROH_frequency_fetal', input)])
freqf= read.table(freqf, h=F, sep= '\t')
names(freqf)= c('chr','pos','freqf')

flist= unlist(input[grep('/ROH/harvest/results/maps_cox/maternal/', input)])
res_m= lapply(flist, read.table)
res_m= do.call("rbind", res_m)
res_m= data.frame(res_m)

flist= unlist(input[grep('/mnt/work/pol/ROH/harvest/results/maps_cox/fetal/', input)])
res_f= lapply(flist, read.table)
res_f= do.call("rbind", res_f)
res_f= data.frame(res_f)

freq= inner_join(freqm, freqf, by= c('chr','pos'))

}

if (!grepl('harvest', input[1])){

moms= unlist(input[grep('runs_mfr_maternal.txt', input)])
moms= read.table(moms, h=T, sep= '\t')
dads= unlist(input[grep('runs_mfr_paternal.txt', input)])
dads= read.table(dads, h=T, sep= '\t')
fets= unlist(input[grep('runs_mfr_fetal.txt', input)])
fets= read.table(fets, h=T, sep= '\t')
q1= unlist(input[grep('q1_pdb315_v9.csv', input)])
q1= read.table(q1, h=T, sep= '\t')

fmoms= unlist(input[grep('/results/het/maternal', input)])
fmoms= fread(fmoms)

fdads= unlist(input[grep('/results/het/paternal', input)])
fdads= fread(fdads)

ffets= unlist(input[grep('/results/het/fetal', input)])
ffets= fread(ffets)

moms= full_join(moms, fmoms, by= 'IID')
dads= full_join(dads, fdads, by= 'IID')
fets= full_join(fets, ffets, by= 'IID')

filter_df= function(df){
df= mutate(df, spont= as.numeric(FSTART=='Spontan' & ((KSNITT=='') | KSNITT== 'Uspesifisert' | KSNITT== 'Akutt keisersnitt') &
                INDUKSJON_PROSTAGLANDIN=='Nei' &
                INDUKSJON_ANNET=='Nei' &
                INDUKSJON_OXYTOCIN=='Nei' &
                INDUKSJON_AMNIOTOMI=='Nei'),
                PARITY0= as.numeric(PARITET_5=='0 (førstegangsfødende)'))
names(df)[names(df) == 'SentrixID'] <- 'SentrixID_1'
return(df)
}

q1$AA1124= as.integer(q1$AA1124)
q1$AA1125= as.integer(q1$AA1125)
q1$AA1128= as.integer(q1$AA1128)
q1$AA1129= as.integer(q1$AA1129)

q1$edu[q1$AA1124 <3 | q1$AA1125<3]= 1
q1$edu[q1$AA1124 ==3 | q1$AA1125==3]= 2
q1$edu[q1$AA1124 ==4 | q1$AA1125==4]= 2
q1$edu[q1$AA1124 ==5 | q1$AA1125==5]= 3
q1$edu[q1$AA1124 ==6 | q1$AA1125==6]= 4
q1$edu[(is.na(q1$edu) & q1$AA1128==6) | (is.na(q1$edu) & q1$AA1129==1)]= 5

q1$AA1315= as.integer(q1$AA1315)
q1$AA1316= as.integer(q1$AA1316)

q1$AA1315[q1$AA1315==1 | q1$AA1315==2| q1$AA1315==3| q1$AA1315==4]= 0
q1$AA1315[q1$AA1315==5 | q1$AA1315==6 | q1$AA1315==7]= 1
q1$AA1316[q1$AA1316==1 | q1$AA1316==2| q1$AA1316==3| q1$AA1316==4]= 0
q1$AA1316[q1$AA1316==5 | q1$AA1316==6 | q1$AA1316==7]= 1

q1$income[q1$AA1315==0 & q1$AA1316==0]= 0
q1$income[q1$AA1315==0 & is.na(q1$AA1316)]= 0
q1$income[is.na(q1$AA1315) & q1$AA1316==0]= 0
q1$income[q1$AA1315==1 | q1$AA1316==1]= 1
q1$income[q1$AA1315==1 & is.na(q1$AA1316)]= 1
q1$income[is.na(q1$AA1315) & q1$AA1316==1]= 1
q1$income[q1$AA1315==1 & q1$AA1316==1]= 2

q1$height= ifelse(q1$AA87>= 140 | q1$AA87<= 210, q1$AA87, NA)
q1= q1 %>% select(PREG_ID_315, edu, income, height)
q1$edu= as.factor(q1$edu)

m= unlist(input[grep('_maternal.hom', input)])
m= fread(m)

ms= m %>% group_by(CHR) %>% summarize(n= n(), KB= median(KB, na.rm=T))
d= unlist(input[grep('_paternal.hom', input)])
d= fread(d)

ds= d %>% group_by(CHR) %>% summarize(n= n(), KB= median(KB, na.rm=T))

fe= unlist(input[grep('_fetal.hom', input)])
fe= fread(fe)

fes= fe %>% group_by(CHR) %>% summarize(n= n(), KB= median(KB, na.rm=T))

moms= filter_df(moms)
dads= filter_df(dads)
fets= filter_df(fets)

moms_c= inner_join(moms, q1, by= 'PREG_ID_315')
dads_c= inner_join(dads, q1, by= 'PREG_ID_315')
fets_c= inner_join(fets, q1, by= 'PREG_ID_315')

freqm= unlist(input[grep('/runs/frequency/ROH_frequency_maternal', input)])
freqm= read.table(freqm, h=F, sep= '\t')
names(freqm)= c('chr','pos','freqm')

freqd= unlist(input[grep('/runs/frequency/ROH_frequency_paternal', input)])
freqd= read.table(freqd, h=F, sep= '\t')
names(freqd)= c('chr','pos','freqd')

freqf= unlist(input[grep('runs/frequency/ROH_frequency_fetal', input)])
freqf= read.table(freqf, h=F, sep= '\t')
names(freqf)= c('chr','pos','freqf')



freq= inner_join(freqm, freqf, by= c('chr','pos'))

flist_m= unlist(input[grep('/results/maps_cox/maternal/', input)])
res_m= lapply(flist_m, read.table)
res_m= do.call("rbind", res_m)
res_m= data.frame(res_m)

flist_f= unlist(input[grep('/results/maps_cox/fetal/', input)])
res_f= lapply(flist_f, read.table)
res_f= do.call("rbind", res_f)
res_f= data.frame(res_f)

}

bin_funk= function(x){
x= na.omit(x)
bin= diff(range(x)) / (2 * IQR(x) / length(x)^(1/3))
return(bin)
}

bim= unlist(input[grep('runs/maternal_input_ROH_geno.txt', input)])
bim= read.table(bim)
bim= as.character(bim[2,1])
bim= fread(bim)
names(bim)= c('CHR', 'SNP', 'POS', 'cM', 'A1', 'A2')

arg= unlist(input[grep('/mnt/work/pol/ROH/arguments/arg_R2', input)])
arg= fread(arg)
arg$fname= lapply(arg$file, function(x) unlist(strsplit(x, '/'))[8])
arg$pruning= ifelse(grepl('hard', arg$fname), 3, ifelse(grepl('moderate', arg$fname), 2, ifelse(grepl('soft', arg$fname), 1,0)))
arg$bp_cm= ifelse(grepl('_bp', arg$fname), 'BP', 'cM')

arg$fname= gsub('.*fetal_', '',arg$fname)
arg$fname= gsub('1e-07', '0.0000001', arg$fname)
arg$fname= gsub('.hom.indiv', '',arg$fname)
arg= arg %>% separate(fname, c('dens', 'SNP', 'length', 'het', 'GAP'), sep= '_')


maxarg= unlist(input[grep('/mnt/work/pol/ROH/arguments/max_R2', input)])
maxarg= fread(maxarg)
maxarg$flags= c('Pruning', 'density', 'SNP', 'length', 'het', 'GAP')

nSNP= as.numeric(maxarg[3,1])
pvalue_t= (0.05 / nrow(bim)) * nSNP


```


# **Aims**
1. To describe ROH patterns in trios,
2. to investigate whether maternal, paternal and fetal ROH and
3. specific ROH areas of the genome are associated with spontaneous delivery.


# **Methods**

## Genotype

Analysis was performed in two Norwegian data sets (HARVEST and ROTTERDAM1) linked to the Norwegian Medical Birth Register.

HARVEST samples (30000 trios) were genotyped using two different Illumina-Chip Array (HumanCoreExome12 and HumanCoreExome24), and ROTTERDAM1 samples using Illumina’s Global Screening Array v.1.0. Sample size specification, see below. 

Similar QC was applied to both datasets. Samples with major ethnicity other than CEU or cryptic relatedness to other samples were excluded (somwehere closer than second-third cousins).

## Phenotype

We included one pregnancy per mother and only **singleton pregnancies** with a **live birth** and a gestational age between 150 and 308 days. Artifitially conceived pregnancies, abruptio placentae, placenta previa, polyhydramnion and fetal malformations were excluded.

## ROH Calls

Despite some researchers giving advice on how to best tune ROH calling parameters, the definition remains arbitrary. Here, we were interested in ROH that were IBD, which led us to think that parental IBD should be correlated to ROH. To we leveraged the relationship between that should exist between autozygosity and identical-by-descent. In this scenario, we identified the parameters by maximizing the R2 between ROH and parental IBD. 
Our interest in ROH lays on its straight connection with autozygozity. 

All analysis were carried out in separate for each data set and sample (maternal, paternal and fetal) for autosomal chromosomes. PLINK was used for ROH calling.  
Main exposure is total length of ROH (sum of bp for all ROH within individual)- scaled and centered.  


## Analysis

The effect of ROH on time-to-spontaneous delivery was estimated using Cox proportional hazards. Gestational age was defined as the time (either by UL - HARVEST or LMP - HUNT), and spontaneous delivery as event. C-sections and inductions were censored.


$$h= h_{0}(t)e^{\sum_{i=1}^{n} \beta_{i}X_{i}}$$

# **ROH Calling arguments and it's relation to IBD**

```{r echo=F, dpi= 300}
argm= gather(arg, arguments, value, dens:bp_cm, factor_key= TRUE)

z= ggplot(argm %>% filter(arguments== 'bp_cm'), aes(x= value, y= R2))  +
geom_boxplot() +
theme_bw(base_family = "Source Sans Pro") +
ylim(0, 1) +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Base pairs or centiMorgans') +
    ylab('R2') +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))


x= ggplot(argm %>% filter(arguments== 'dens'), aes(x= value, y= R2))  + 
geom_boxplot() + 
theme_bw(base_family = "Source Sans Pro") +
ylim(0, 1) +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('SNP density') +
    ylab('R2') +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))


x1= ggplot(argm %>% filter(arguments== 'SNP'), aes(x= as.factor(as.numeric(value)), y= R2))  +
geom_boxplot() +
theme_bw(base_family = "Source Sans Pro") +
ylim(0, 1) +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('N SNPs') +
    ylab('R2') +
   theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))
 

x2= ggplot(argm %>% filter(arguments== 'het'), aes(x= value, y= R2))  +
geom_boxplot() +
theme_bw(base_family = "Source Sans Pro") +
ylim(0, 1) +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('N heterozygous allowed') +
    ylab('R2') +
   theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))



x3= ggplot(argm %>% filter(arguments== 'pruning'), aes(x= value, y= R2))  +
geom_boxplot() +
theme_bw(base_family = "Source Sans Pro") +
ylim(0, 1) +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Pruning') +
    ylab('R2') +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

lay= rbind(c(0,0),
	c(1, 2),
	c(3, 4),
	c(5, 5))

grid.arrange(grobs= list(x3, x, x2, z, x1), layout_matrix= lay)


#tidy(lm(R2~ dens + SNP + het + GAP + pruning, arg)) %>% kable(caption= 'Associations between ROH calling parameters and R2 (ROH against parental IBD).', digits= 3, )


R2= summary(lm(R2~ dens + SNP + het + GAP + pruning + bp_cm, arg))$r.squared


```

R2 with a model containing density, heterozygosity, number of SNPs, pruning and the gap between adjacent ROH gives an R2 of: `r round(R2,3)`

```{r echo=F, dpi= 300}

ggplot(arg, aes(x= as.factor(as.numeric(SNP)), y= R2))  +
geom_boxplot() +
facet_grid(rows= vars(pruning)) +
theme_bw(base_family = "Source Sans Pro") +
ylim(0, 1) +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank(),
	   panel.background = element_rect(fill = NA, color = "black")
    ) + xlab('Number of SNPs') +
    ylab('R2') +
 ggtitle('Each block represents a pruning category: 0-3, none-soft-moderate-hard') + 
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2)) +
	geom_hline(yintercept= max(arg$R2), color= 'coral', size= 0.2, linetype='dashed' )

R2= summary(lm(R2~  SNP * as.factor(pruning), arg))$r.squared

```

R2 with a model containing the number of SNPs, pruning and the their interaction gives an R2 of: `r round(R2,3)`

# **Final parameters for ROH calling**

Thus, the **final parameters** were as follows:  
    - **Pruning:** `r as.numeric(maxarg[1,1])`  
    - **SNP density (--homozyg-density):** `r as.numeric(maxarg[2,1])`  
    - **Number of SNPs (--homozyg-snp):** `r as.numeric(maxarg[3,1])`  
    - **Number of heterozygous SNPs allowed (--homozyg-het):** `r as.numeric(maxarg[5,1])`  
    - Gap between adjacent ROH (--homozyg-gap): `r as.numeric(maxarg[6,1])`  

ROH length was always set to a minimum of 0.0000001.  



# **Runs of Homozygosity (ROH) description**

## ROH distribution by chromosome

```{r echo=F, dpi= 300}


x= ggplot(m, aes(factor(CHR), KB, fill= factor(CHR)))+
geom_violin() +
scale_fill_manual(values = rep(c('#a1dab4', '#41b6c4'), 11)) +
ylim(0, 5000) +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Chromosome') +
    ylab('cM') +
ggtitle("Mothers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x1=  ggplot(d, aes(factor(CHR), KB, fill= factor(CHR)))+
geom_violin() +
scale_fill_manual(values = rep(c('#a1dab4', '#41b6c4'), 11)) +
ylim(0, 5000) +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Chromosome') +
    ylab('cM') +
ggtitle("Fathers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x2= ggplot(fe, aes(factor(CHR), KB, fill= factor(CHR)))+
geom_violin() +
scale_fill_manual(values = rep(c('#a1dab4', '#41b6c4'), 11)) +
ylim(0, 5000) +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Chromosome') +
    ylab('cM') +
ggtitle("Children") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))


grid.arrange(x, x1, x2)
```

## Length of ROH per chromosome

```{r echo=F, dpi= 300}
x= ggplot(ms, aes(factor(CHR), KB, color= n)) +
geom_point() +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Chromosome') +
    ylab('Median (Kb)') +
ggtitle("Mothers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x1= ggplot(ds, aes(factor(CHR), KB, color= n)) +
geom_point() +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Chromosome') +
    ylab('Median (Kb)') +
ggtitle("Fathers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x2= ggplot(fes, aes(factor(CHR), KB, color= n )) +
geom_point() +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Chromosome') +
    ylab('Median (Kb)') +
ggtitle("Children") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

grid.arrange(x, x1, x2)

```

## Total number of ROH per chromosome

```{r echo=F, dpi= 300}

x= ggplot(ms, aes(factor(CHR), n, color= KB)) +
geom_point() +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Chromosome') +
    ylab('N ROH') +
ggtitle("Mothers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x1= ggplot(ds, aes(factor(CHR), n, color= KB)) +
geom_point() +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Chromosome') +
    ylab('N ROH') +
ggtitle("Fathers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x2= ggplot(fes, aes(factor(CHR), n, color= KB)) +
geom_point() +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Chromosome') +
    ylab('N ROH') +
ggtitle("Children") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

grid.arrange(x, x1, x2)

```


---


Total number of:   
|	- mothers: `r nrow(moms)`  
|	- fathers: `r nrow(dads)`  
|	- offspring: `r nrow(fets)`  

```{r filter_data_sets, echo= F, message= F, dpi= 300}

x= moms %>% summarize(mKB= mean(KB, na.rm=T), mNSEG= mean(NSEG, na.rm=T), mFKB= mean(FKB, na.rm=T))
x1= dads %>% summarize(mKB= mean(KB, na.rm=T), mNSEG= mean(NSEG, na.rm=T), mFKB= mean(FKB, na.rm=T))
x2= fets %>% summarize(mKB= mean(KB, na.rm=T), mNSEG= mean(NSEG, na.rm=T), mFKB= mean(FKB, na.rm=T))

d= do.call("rbind", list(x, x2, x1))
d$role= c('Mother', 'Father', 'Child')

#print(paste('Total number of mothers:', nrow(moms)))
#print(paste('Total number of fathers:', nrow(dads)))
#print(paste('Total number of children:', nrow(fets)))

d= d %>% select(role, mKB, mNSEG, mFKB)

kable(d, col.names = c('Role', 'ROH length', 'N ROH', 'F_ROH'), caption= 'ROH description in mother, father and offspring', digits= 1, align= c('l','c','c', 'c')) %>% kable_styling(bootstrap_options = c("striped", "hover"), full_width = F, position = "center", font_size = 12) %>%   column_spec(1, bold = T, border_right = T)

```

## Total FROH length distribution

```{r echo=F, dpi= 300}
bins= bin_funk(moms$FKB)

ggplot(moms, aes(FKB)) +
geom_histogram(aes(y=..density..) , bins = bins, colour="#41b6c4", fill="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Total FROH length') +
    ylab('Count') +
ggtitle("Mothers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

bins= bin_funk(dads$FKB)


```

## Correlations between number of ROH and total ROH length

```{r echo=F, dpi= 300}
ggplot(moms, aes(NSEG, FKB)) +
geom_point(alpha= 0.5, colour= '#41b6c4') +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme(text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Number of ROH') +
    ylab('Total FROH length') +
ggtitle("Mothers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

```

## ROH and PCA

```{r echo=F, dpi= 300}

ggplot(moms, aes(PC1, FKB, color= PC1)) +
geom_point() +
geom_smooth(method = "lm", se = FALSE, colour="#2c7fb8") +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('PC1') +
    ylab('Length of FROHs') +
ggtitle("Mothers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))
```

# Results - ROH and spontaneous delivery


---


```{r echo=F, dpi= 300}

crudecoxm= coxph(Surv(SVLEN_UL_DG, spont)~ FKB, moms)
crudecoxd= coxph(Surv(SVLEN_UL_DG, spont)~ FKB, dads)
crudecoxf= coxph(Surv(SVLEN_UL_DG, spont)~ FKB, fets)

minadjcoxm= coxph(Surv(SVLEN_UL_DG, spont)~ FKB + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0, moms)
minadjcoxd= coxph(Surv(SVLEN_UL_DG, spont)~ FKB + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0, dads)
minadjcoxf= coxph(Surv(SVLEN_UL_DG, spont)~ FKB + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0, fets)

fullcoxm= coxph(Surv(SVLEN_UL_DG, spont)~ FKB + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0 + edu, moms_c)
fullcoxd= coxph(Surv(SVLEN_UL_DG, spont)~ FKB + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0 + edu, dads_c)
fullcoxf= coxph(Surv(SVLEN_UL_DG, spont)~ FKB + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0 + edu, fets_c)

x= summary(crudecoxm)$conf.int[1, c(1,3,4)]
x1= summary(minadjcoxm)$conf.int[1, c(1,3,4)]
x2= summary(fullcoxm)$conf.int[1, c(1,3,4)]

d= do.call("rbind", list(x, x1, x2))
d= data.frame(d)
d$role= c('Mother', 'Mother', 'Mother')
d$model= c('Crude', 'Model 1', 'Model 2')

x= summary(crudecoxd)$conf.int[1, c(1,3,4)]
x1= summary(minadjcoxd)$conf.int[1, c(1,3,4)]
x2= summary(fullcoxd)$conf.int[1, c(1,3,4)]

d1= do.call("rbind", list(x, x1, x2))
d1= data.frame(d1)
d1$role= c('Father', 'Father', 'Father')
d1$model= c('Crude', 'Model 1', 'Model 2')

x= summary(crudecoxf)$conf.int[1, c(1,3,4)]
x1= summary(minadjcoxf)$conf.int[1, c(1,3,4)]
x2= summary(fullcoxf)$conf.int[1, c(1,3,4)]

d2= do.call("rbind", list(x, x1, x2))
d2= data.frame(d2)
d2$role= c('Fetal', 'Fetal', 'Fetal')
d2$model= c('Crude', 'Model 1', 'Model 2')
d= do.call("rbind", list(d,d1,d2))
d= data.frame(d)

d= d %>% select(role, model, exp.coef., lower..95, upper..95)

kable(d, caption= 'FROH length and risk of spontaneous delivery', col.names = c('Role', 'Model', 'Hazard Ratio', 'Lower 95%CI', 'Upper 95%CI'), digits= 3) %>% kable_styling(bootstrap_options = c("striped", "hover"), full_width = F, position = "center", font_size = 12) %>%   column_spec(1, bold = T, border_right = T)  %>%   footnote(general = 'Adjusted by: crude: none; model 1: PC1-10 and parity; model 2: model 1 + maternal educational attainment and household income.')
```


```{r echo=F, dpi= 300}
ggplot(d %>% filter(model== 'Model 2'), aes(x= role, y= exp.coef.)) +
geom_errorbar(aes(ymin= lower..95 , ymax= upper..95), width= 0.1) +
geom_point() +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme(text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()) +
xlab('Sample') +
    ylab('HR (95% CI)') +
ggtitle("FROH and spontaneous delivery risk (model 2)") +
geom_hline(aes(yintercept= 1), colour= 'grey', linetype= 'dashed') +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))
```


## Maternal FROH and spontaneous delivery

```{r echo=F, fig.width= 12, fig.height= 10, dpi= 300}



moms= moms %>% mutate(QKB= ifelse(FKB==0,0,1))
coxm= survfit(Surv(SVLEN_UL_DG, spont)~ QKB, moms)

p= ggsurvplot(
  coxm,                     # survfit object with calculated statistics.
  data = moms,  # data used to fit survival curves. 
  risk.table = TRUE,
  palette =
    c( "#41b6c4","#253494"),# show risk table. "gold"
  pval = TRUE,             # show p-value of log-rank test.
  conf.int = FALSE,         # show confidence intervals for 
  # point estimaes of survival curves.
  xlim = c(250,300),        # present narrower X axis, but not affect
  # survival estimates.
  break.time.by = 10,     # break X axis in time intervals by 500.
  ggtheme = theme_minimal(base_family ='URWHelvetica'), # customize plot and risk table with a theme.
  risk.table.y.text.col = T, # colour risk table text annotations.
  risk.table.y.text = FALSE, # show bars instead of names in text annotations
  # in legend of risk table
  xlab = "Time in days",
  legend.labs = c("No ROH", "ROH"),
  surv.median.line = "hv",
  censor=FALSE,
  font.family= 'URWHelvetica')

p= ggpar(p,  font.title    = c(16, "bold", "#969696"),
  font.subtitle = c(16, "bold.italic", "purple"),
  font.x        = c(16, "plain", "black"),
  font.y        = c(16, "plain", "black"),
  font.xtickslab = c(14, "plain", "#969696"),
  #font.ytickslab = c(12, "plain", "#969696"),
  legend = "top"
)

p
```




---

```{r echo=F, eval=F, dpi= 300}
fullcoxml= coxph(Surv(SVLEN_UL_DG, spont)~ scale(KB_long) + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0, moms)
fullcoxdl= coxph(Surv(SVLEN_UL_DG, spont)~ scale(KB_long) + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0, dads)
fullcoxfl= coxph(Surv(SVLEN_UL_DG, spont)~ scale(KB_long) + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0, fets)

x= summary(fullcoxml)$conf.int[1, c(1,3,4)]
x1= summary(fullcoxdl)$conf.int[1, c(1,3,4)]
x2= summary(fullcoxfl)$conf.int[1, c(1,3,4)]

d= do.call("rbind", list(x, x1, x2))
d= data.frame(d)
d$role= c('Mother', 'Father', 'Child')
d$bp= 'Long ROH'


fullcoxms= coxph(Surv(SVLEN_UL_DG, spont)~ scale(KB_short) + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0, moms)
fullcoxds= coxph(Surv(SVLEN_UL_DG, spont)~ scale(KB_short) + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0, dads)
fullcoxfs= coxph(Surv(SVLEN_UL_DG, spont)~ scale(KB_short) + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0, fets)

x= summary(fullcoxms)$conf.int[1, c(1,3,4)]
x1= summary(fullcoxds)$conf.int[1, c(1,3,4)]
x2= summary(fullcoxfs)$conf.int[1, c(1,3,4)]

ds= do.call("rbind", list(x, x1, x2))
ds= data.frame(ds)
ds$role= c('Mother', 'Father', 'Child')
ds$bp= 'Short ROH'
ds= ds %>% select(role, bp, exp.coef., lower..95, upper..95)

d= rbind(d, ds)

ggplot(d, aes(x= bp, y= exp.coef.)) +
facet_grid(. ~ role) +
geom_errorbar(aes(ymin= lower..95 , ymax= upper..95), width= 0.1) +
geom_point() +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme(text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank(),
        strip.background =element_rect(fill="white")) +
xlab('Sample') +
    ylab('HR for spontaneous delivery (95% CI)') +
ggtitle("ROH and Spontaneous delivery risk") +
geom_hline(aes(yintercept= 1), colour= 'grey', linetype= 'dashed') +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))
```

# Mapping of ROHs


---


## Frequency of ROHs in mothers and children

```{r echo=F, dpi= 300}


ggplot(freq, aes(x= freqm, y= freqf)) +
geom_point(alpha= 0.4, size= 0.01) +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme(text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank(),
        strip.background =element_rect(fill="white")) +
xlab('Maternal ROH frequency') +
    ylab('Offspring ROH frequency') +
ggtitle("Per-SNP ROH frequency in mother-child pairs") +
geom_abline(intercept= 0, slope=1, colour= 'grey', linetype= 'dashed') +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))
```

## Maternal ROH mapping and spontaneous delivery


```{r echo=F, dpi= 300}

names(freqm)=  c('chr','pos','freq')

mhtf= function(df){ # Column names needed (chr, pos, freq)
  if (class(df$chr) != 'integer') {
    df$chr= ifelse(df$chr== 'X', '23', df$chr)
    df$chr= as.numeric(as.character(df$chr))
  }
  require(dplyr)
  require(ggplot2)
  df = df %>% filter(!is.na(chr))

  don <- df %>%
    group_by(chr)      %>%
    summarise(chr_len= max(pos)) %>%
    mutate(tot= cumsum(chr_len)-chr_len) %>% # Calculate cumulative position of each chromosome
    select(-chr_len) %>%
    left_join(df, ., by= 'chr') %>%
    arrange(chr, pos) %>% # Add a cumulative position of each SNP
    mutate( BPcum=pos+tot)

  axisdf = don %>% group_by(chr) %>% summarize(center=( max(BPcum) + min(BPcum) ) / 2 )
  names(axisdf)= c('chr', 'center')

  ggplot(don, aes(x=BPcum, y= freq)) +
    geom_point( aes(color=as.factor(chr)), alpha= 1/10, size=0.01) +   # Show all points
    scale_colour_manual(values = rep(c("black"),  22 )) + # 08519c 3182bd "#006d2c","#31a354"
    scale_x_continuous( label = axisdf$chr, breaks= axisdf$center, expand=c(0,0) ) + # custom X axis
    scale_y_continuous(expand = c(0, 0), limits= c(0, 0.1)) +     # remove space between plot area and x axis
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") + # Custom the theme
    theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Chromosome') +
    ylab('ROH Rel. frequency') +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))
}

x= mhtf(freqm)

mht= function(df){ # Column names needed (chr, pos, pvalue)
  if (class(df$chr) != 'integer') {
    df$chr= ifelse(df$chr== 'X', '23', df$chr)
    df$chr= as.numeric(as.character(df$chr))
  }
  require(dplyr)
  require(ggplot2)
  df = df %>% filter(!is.na(chr))

  don <- df %>%
    group_by(chr)      %>%
    summarise(chr_len= max(pos)) %>%
    mutate(tot= cumsum(chr_len)-chr_len) %>% # Calculate cumulative position of each chromosome
    select(-chr_len) %>%
    left_join(df, ., by= 'chr') %>%
    arrange(chr, pos) %>% # Add a cumulative position of each SNP
    mutate( BPcum=pos+tot)

  axisdf = don %>% group_by(chr) %>% summarize(center=( max(BPcum) + min(BPcum) ) / 2 )
  names(axisdf)= c('chr', 'center')

  ggplot(don, aes(x=BPcum, y= -log10(pvalue))) +
    geom_point( aes(color=as.factor(chr)), size=0.1) +   # Show all points
    scale_colour_manual(values = rep(c("#1d91c0","#41b6c4"),  22 )) + # 08519c 3182bd "#006d2c","#31a354"
    scale_x_continuous( label = axisdf$chr, breaks= axisdf$center, expand=c(0,0) ) + # custom X axis
    scale_y_continuous(expand = c(0, 0), limits= c(0, max(-log10(don$pvalue)) + 1)) +     # remove space between plot area and x axis
#theme_bw() +
	   theme_bw(base_family = "Source Sans Pro") + # Custom the theme
    theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Chromosome') +
    ylab('-log10(P-value)') +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2)) +
#    geom_hline(yintercept = -log10(1*10^-5), size= 0.5, color= '#253494', alpha= 0.7) +
    geom_hline(yintercept = -log10(pvalue_t), size= 0.5, color= 'gold', alpha= 0.7)

}


names(res_m)= c('id', 'N', 'beta', 'sd', 'pvalue', 'schoenfeld', 'p_scheonfeld')
res_m= res_m %>% separate(id, c('chr', 'pos'), ':')
res_m= res_m %>% mutate(chr= as.numeric(chr), pos= as.numeric(pos))

x1= mht(res_m)

grid.arrange(x1, x, nrow= 2)
```


```{r echo=F, dpi= 300}

beta_mht= function(df){ # Column names needed (chr, pos, pvalue)
 require(dplyr)
  require(ggplot2)
  df = df %>% filter(!is.na(chr))

  don <- df %>%
    group_by(chr)      %>%
    summarise(chr_len= max(pos)) %>%
    mutate(tot= cumsum(chr_len)-chr_len) %>% # Calculate cumulative position of each chromosome
    select(-chr_len) %>%
    left_join(df, ., by= 'chr') %>%
    arrange(chr, pos) %>% # Add a cumulative position of each SNP
    mutate( BPcum=pos+tot)

  axisdf = don %>% group_by(chr) %>% summarize(center=( max(BPcum) + min(BPcum) ) / 2 )
  names(axisdf)= c('chr', 'center')

  ggplot(don, aes(x=BPcum, y= I(beta/sd))) +
    geom_point( aes(color= factor(maf)), size=0.1) +   # Show points
    scale_colour_manual(values = c("#c7e9b4","#41b6c4","#984ea3", "#984ea3")) + # 08519c 3182bd "#006d2c","#31a354"
    scale_x_continuous( label = axisdf$chr, breaks= axisdf$center, expand=c(0,0) ) + # custom X axis
    #scale_y_continuous(expand = c(0, 0), limits= c(0, max((I(don$beta/don$se))) + 1)) +     # remove space between plot area and x axis
#theme_bw() +
    theme_bw(base_family = "Source Sans Pro") + # Custom the theme
    theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Chromosome') +
    ylab('Z-score') +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2)) +
    geom_hline(yintercept = 2.96, size= 0.5, color= '#253494', alpha= 0.7) +
    geom_hline(yintercept = -2.96, size= 0.5, color= '#253494', alpha= 0.7) +
geom_hline(yintercept = 0, size= 0.5, color= 'grey',linetype='dashed', alpha= 0.7)

}
```


```{r echo=F, dpi= 300}

res_m= inner_join(res_m, freqm, by= c('chr', 'pos'))
res_m$maf= ifelse(res_m$freq>=0.5, 3, ifelse(res_m$freq>= 0.01, 2, ifelse(res_m$freq>=0.001, 1, 0)))
beta_mht(res_m)

res_m$Zscore= res_m$beta / res_m$sd

bins= bin_funk(res_m$Zscore)

ggplot(res_m, aes(Zscore)) +
geom_histogram(aes(y=..density..) , bins = bins, colour="#41b6c4", fill="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Effect size (zscore)') +
    ylab('Count') +
ggtitle("Mothers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))


```


## Materno-fetal correlations

```{r echo=F, dpi= 300}


names(res_f)= c('id', 'N_f', 'beta_f', 'sd_f', 'pvalue_f', 'schoenfeld_f', 'p_scheonfeld_f')
res_f= res_f %>% separate(id, c('chr', 'pos'), ':')
res_f= res_f %>% mutate(chr= as.numeric(chr), pos= as.numeric(pos))


df_f= inner_join(res_f, res_m, by= c('chr','pos'))

df_f= df_f %>% filter(pvalue< 1*10^-4)

df_f= group_by(df_f, chr) %>%
               mutate(d=pos-lag(pos, default=-Inf), clumpid=cumsum(d>1e5)) %>%
              group_by(chr, clumpid) %>%
filter(rank(pvalue, ties.method = "random")==1)

```

```{r echo=F, dpi= 300}
ggplot(df_f, aes(x= I(beta/sd_f), y= I(beta_f/sd_f))) +
geom_point(size= 0.01) +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme(text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank(),
        strip.background =element_rect(fill="white")) +
xlab('Maternal ROH beta') +
    ylab('Offspring ROH beta') +
ggtitle("Per-SNP ROH beta estimates") +
#geom_abline(intercept= 0, slope=1, colour= 'grey', linetype= 'dashed') +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))
```



## Fetal ROH and risk of spontaneous delivery   


```{r echo= F, dpi= 300}

names(freqf)= c('chr', 'pos','freq') 
names(res_f)= c('chr','pos','N', 'beta', 'sd', 'pvalue', 'schoenfeld', 'p_scheonfeld') 

x= mhtf(freqf) 
 
x1= mht(res_f) 

grid.arrange(x1, x, nrow= 2)

```

# **Alternative measure of inbreeding: Fhom**

## Fhom distribution  

```{r echo=F, dpi= 300}

bins= bin_funk(moms$none_F)

x= ggplot(moms, aes(none_F)) +
geom_histogram(aes(y=..density..) , bins = bins, colour="#41b6c4", fill="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Excess homozygosity') +
    ylab('Count') +
ggtitle("No pruning") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

bins= bin_funk(moms$soft_F)

x1= ggplot(moms, aes(soft_F)) +
geom_histogram(aes(y=..density..) , bins = bins, colour="#41b6c4", fill="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Excess homozygosity') +
    ylab('Count') +
ggtitle("Soft pruning") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

bins= bin_funk(moms$moderate_F)
x2= ggplot(moms, aes(moderate_F)) +
geom_histogram(aes(y=..density..) , bins = bins, colour="#41b6c4", fill="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Excess homozygosity') +
    ylab('Count') +
ggtitle("Moderate pruning") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

bins= bin_funk(moms$hard_F)
x3= ggplot(moms, aes(hard_F)) +
geom_histogram(aes(y=..density..) , bins = bins, colour="#41b6c4", fill="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Excess homozygosity') +
    ylab('Count') +
ggtitle("Hard pruning") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))


grid.arrange(x, x1, x2, x3, nrow= 2, top= 'Mothers')

```

## Fhom and PCA

```{r echo=F, dpi= 300}

ggplot(moms, aes(PC1, none_F, color= PC1)) +
geom_point() +
geom_smooth(method = "lm", se = FALSE, colour="#2c7fb8") +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('PC1') +
    ylab('Excess homozygosity') +
ggtitle("Mothers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

```


## Fhom and ROH

```{r echo=F, dpi= 300}

x= ggplot(moms, aes(none_F, FKB, color= PC1)) +
geom_point() +
geom_smooth(method = "lm", se = FALSE, colour="#2c7fb8") +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Excess homozygosity') +
    ylab('FROH') +
ggtitle("No pruning") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x1= ggplot(moms, aes(soft_F, FKB, color= PC1)) +
geom_point() +
geom_smooth(method = "lm", se = FALSE, colour="#2c7fb8") +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Excess homozygosity') +
    ylab('FROH') +
ggtitle("Soft pruning") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x2= ggplot(moms, aes(moderate_F, FKB, color= PC1)) +
geom_point() +
geom_smooth(method = "lm", se = FALSE, colour="#2c7fb8") +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Excess homozygosity') +
    ylab('FROH') +
ggtitle("Moderate pruning") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x3= ggplot(moms, aes(hard_F, FKB, color= PC1)) +
geom_point() +
geom_smooth(method = "lm", se = FALSE, colour="#2c7fb8") +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Excess homozygosity') +
    ylab('FROH') +
ggtitle("Hard pruning") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

grid.arrange(x, x1, x2, x3, nrow= 2, top= 'Mothers')


x= ggplot(dads, aes(none_F, FKB, color= PC1)) +
geom_point() +
geom_smooth(method = "lm", se = FALSE, colour="#2c7fb8") +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Excess homozygosity') +
    ylab('FROH') +
ggtitle("No pruning") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x1= ggplot(dads, aes(soft_F, FKB, color= PC1)) +
geom_point() +
geom_smooth(method = "lm", se = FALSE, colour="#2c7fb8") +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Excess homozygosity') +
    ylab('FROH') +
ggtitle("Soft pruning") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x2= ggplot(dads, aes(moderate_F, FKB, color= PC1)) +
geom_point() +
geom_smooth(method = "lm", se = FALSE, colour="#2c7fb8") +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Excess homozygosity') +
    ylab('FROH') +
ggtitle("Moderate pruning") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x3= ggplot(dads, aes(hard_F, FKB, color= PC1)) +
geom_point() +
geom_smooth(method = "lm", se = FALSE, colour="#2c7fb8") +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Excess homozygosity') +
    ylab('FROH') +
ggtitle("Hard pruning") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

grid.arrange(x, x1, x2, x3, nrow= 2, top= 'Fathers')

```

## Results - Fhom and spontaneous delivery risk (no pruning)

```{r echo=F, dpi= 300}

crudecoxm= coxph(Surv(SVLEN_UL_DG, spont)~ none_F, moms)
crudecoxd= coxph(Surv(SVLEN_UL_DG, spont)~ none_F, dads)
crudecoxf= coxph(Surv(SVLEN_UL_DG, spont)~ none_F, fets)

minadjcoxm= coxph(Surv(SVLEN_UL_DG, spont)~ none_F + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0, moms)
minadjcoxd= coxph(Surv(SVLEN_UL_DG, spont)~ none_F + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0, dads)
minadjcoxf= coxph(Surv(SVLEN_UL_DG, spont)~ none_F + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0, fets)

fullcoxm= coxph(Surv(SVLEN_UL_DG, spont)~ none_F + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0 + edu, moms_c)
fullcoxd= coxph(Surv(SVLEN_UL_DG, spont)~ none_F + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0 + edu, dads_c)
fullcoxf= coxph(Surv(SVLEN_UL_DG, spont)~ none_F + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0 + edu, fets_c)

x= summary(crudecoxm)$conf.int[1, c(1,3,4)]
x1= summary(minadjcoxm)$conf.int[1, c(1,3,4)]
x2= summary(fullcoxm)$conf.int[1, c(1,3,4)]

d= do.call("rbind", list(x, x1, x2))
d= data.frame(d)
d$role= c('Mother', 'Mother', 'Mother')
d$model= c('Crude', 'Model 1', 'Model 2')

x= summary(crudecoxd)$conf.int[1, c(1,3,4)]
x1= summary(minadjcoxd)$conf.int[1, c(1,3,4)]
x2= summary(fullcoxd)$conf.int[1, c(1,3,4)]

d1= do.call("rbind", list(x, x1, x2))
d1= data.frame(d1)
d1$role= c('Father', 'Father', 'Father')
d1$model= c('Crude', 'Model 1', 'Model 2')

x= summary(crudecoxf)$conf.int[1, c(1,3,4)]
x1= summary(minadjcoxf)$conf.int[1, c(1,3,4)]
x2= summary(fullcoxf)$conf.int[1, c(1,3,4)]

d2= do.call("rbind", list(x, x1, x2))
d2= data.frame(d2)
d2$role= c('Fetal', 'Fetal', 'Fetal')
d2$model= c('Crude', 'Model 1', 'Model 2')
d= do.call("rbind", list(d,d1,d2))
d= data.frame(d)

d= d %>% select(role, model, exp.coef., lower..95, upper..95)

kable(d, caption= 'Fhom (no pruning) and risk of spontaneous delivery', col.names = c('Role', 'Model', 'Hazard Ratio', 'Lower 95%CI', 'Upper 95%CI'), digits= 3) %>% kable_styling(bootstrap_options = c("striped", "hover"), full_width = F, position = "center", font_size = 12) %>%   column_spec(1, bold = T, border_right = T)  %>%   footnote(general = 'Adjusted by: crude: none; model 1: PC1-10 and parity; model 2: model 1 + maternal educational attainment and household income.')


fullcoxm= coxph(Surv(SVLEN_UL_DG, spont)~ none_F + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0 + edu, moms_c)
fullcoxm1= coxph(Surv(SVLEN_UL_DG, spont)~ soft_F + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0 + edu, moms_c)
fullcoxm2= coxph(Surv(SVLEN_UL_DG, spont)~ moderate_F + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0 + edu, moms_c)
fullcoxm3= coxph(Surv(SVLEN_UL_DG, spont)~ hard_F + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0 + edu, moms_c)

x= summary(fullcoxm)$conf.int[1, c(1,3,4)]
x1= summary(fullcoxm1)$conf.int[1, c(1,3,4)]
x2= summary(fullcoxm2)$conf.int[1, c(1,3,4)]
x3= summary(fullcoxm3)$conf.int[1, c(1,3,4)]

d= do.call("rbind", list(x, x1, x2, x3))
d= data.frame(d)
d$pruning= c('None', 'Soft', 'Moderate', 'Hard')

z= ggplot(d, aes(x= pruning, y= exp.coef.)) +
geom_errorbar(aes(ymin= lower..95 , ymax= upper..95), width= 0.1) +
geom_point() +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme(text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()) +
xlab('Pruning mode') +
    ylab('HR (95% CI)') +
ggtitle("Mothers") +
geom_hline(aes(yintercept= 1), colour= 'grey', linetype= 'dashed') +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

fullcoxd= coxph(Surv(SVLEN_UL_DG, spont)~ none_F + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0 + edu, dads_c)
fullcoxd1= coxph(Surv(SVLEN_UL_DG, spont)~ soft_F + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0 + edu, dads_c)
fullcoxd2= coxph(Surv(SVLEN_UL_DG, spont)~ moderate_F + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0 + edu, dads_c)
fullcoxd3= coxph(Surv(SVLEN_UL_DG, spont)~ hard_F + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0 + edu, dads_c)

x= summary(fullcoxd)$conf.int[1, c(1,3,4)]
x1= summary(fullcoxd1)$conf.int[1, c(1,3,4)]
x2= summary(fullcoxd2)$conf.int[1, c(1,3,4)]
x3= summary(fullcoxd3)$conf.int[1, c(1,3,4)]

d= do.call("rbind", list(x, x1, x2, x3))
d= data.frame(d)
d$pruning= c('None', 'Soft', 'Moderate', 'Hard')

z1= ggplot(d, aes(x= pruning, y= exp.coef.)) +
geom_errorbar(aes(ymin= lower..95 , ymax= upper..95), width= 0.1) +
geom_point() +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme(text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()) +
xlab('Pruning mode') +
    ylab('HR (95% CI)') +
ggtitle("Fathers") +
geom_hline(aes(yintercept= 1), colour= 'grey', linetype= 'dashed') +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

fullcoxf= coxph(Surv(SVLEN_UL_DG, spont)~ none_F + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0 + edu, fets_c)
fullcoxf1= coxph(Surv(SVLEN_UL_DG, spont)~ soft_F + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0 + edu, fets_c)
fullcoxf2= coxph(Surv(SVLEN_UL_DG, spont)~ moderate_F + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0 + edu, fets_c)
fullcoxf3= coxph(Surv(SVLEN_UL_DG, spont)~ hard_F + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0 + edu, fets_c)

x= summary(fullcoxf)$conf.int[1, c(1,3,4)]
x1= summary(fullcoxf1)$conf.int[1, c(1,3,4)]
x2= summary(fullcoxf2)$conf.int[1, c(1,3,4)]
x3= summary(fullcoxf3)$conf.int[1, c(1,3,4)]

d= do.call("rbind", list(x, x1, x2, x3))
d= data.frame(d)
d$pruning= c('None', 'Soft', 'Moderate', 'Hard')

z2= ggplot(d, aes(x= pruning, y= exp.coef.)) +
geom_errorbar(aes(ymin= lower..95 , ymax= upper..95), width= 0.1) +
geom_point() +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme(text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()) +
xlab('Pruning mode') +
    ylab('HR (95% CI)') +
ggtitle("Offspring") +
geom_hline(aes(yintercept= 1), colour= 'grey', linetype= 'dashed') +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

grid.arrange(z,z1,z2, ncol= 2, top= 'Fhom and spontaneous delivery risk')
```

# Associations between fetal Fhom and parental IBD

```{r echo=F}

ibd= unlist(input[grep('parental_ibd.txt', input)])
ibd= fread(ibd)
trios= unlist(input[grep('trios.txt', input)])
trios= fread(trios)
ibd= right_join(ibd, fets, by= c('Child'= 'IID'))
ibd$cM= ifelse(is.na(ibd$cM), 0, ibd$cM)


R2= round(summary(lm(none_F~ cM, ibd))$r.squared , 3)
z= ggplot(ibd, aes(x= cM, y= none_F)) +
geom_point() +
geom_smooth(method='lm',formula=y~x) +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme(text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()) +
xlab('Parental total IBD (cM)') +
    ylab('Excess homozygosity') +
ggtitle("No pruning") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2)) +
annotate('text', label= paste("italic(R) ^ 2 ==", R2), x = 400, y = -0.1,
parse=TRUE)

R2= round(summary(lm(soft_F~ cM, ibd))$r.squared, 3)
z1= ggplot(ibd, aes(x= cM, y= soft_F)) +
geom_point() +
geom_smooth(method='lm',formula=y~x) +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme(text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()) +
xlab('Parental total IBD (cM)') +
    ylab('Excess homozygosity') +
ggtitle("Soft pruning") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))+
annotate('text', label= paste("italic(R) ^ 2 ==", R2), x = 400, y = -0.1,
parse=TRUE)

R2=round( summary(lm(moderate_F~ cM, ibd))$r.squared,3)
z2= ggplot(ibd, aes(x= cM, y= moderate_F)) +
geom_point() +
geom_smooth(method='lm',formula=y~x) +
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme(text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()) +
xlab('Parental total IBD (cM)') +
    ylab('Excess homozygosity') +
ggtitle("Moderate pruning") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))+
annotate('text', label= paste("italic(R) ^ 2 ==", R2), x = 400, y = -0.1,
parse=TRUE)

R2= round(summary(lm(hard_F~ cM, ibd))$r.squared,3)
z3= ggplot(ibd, aes(x= cM, y= hard_F)) +
geom_point() +
geom_smooth(method='lm',formula=y~x) + 
#theme_bw() +
theme_bw(base_family = "Source Sans Pro") +
theme(text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()) +
xlab('Parental total IBD (cM)') +
    ylab('Excess homozygosity') +
ggtitle("Hard pruning") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))+
annotate('text', label= paste("italic(R) ^ 2 ==", R2), x = 400, y = -0.1,
parse=TRUE)


grid.arrange(z, z1,z2, z3, ncol= 2, top= 'Parental IBD and offspring Fhom')

```
