---
title: "**Autozigosity and gestional duration: HARVEST report.**"
author:
        - "Pol Sole-Navais"
date: "`r format(Sys.time(), '%d %B, %Y')`"
params:
    rmd: "hunt_report.Rmd"
    maternal: "/mnt/work/pol/ROH/pheno/runs_mfr_maternal.txt"
    paternal: "/mnt/work/pol/ROH/pheno/runs_mfr_paternal.txt"
    fetal: "/mnt/work/pol/ROH/pheno/runs_mfr_fetal.txt"
output:
    html_document:
        highlight: tango
        number_sections: no
        theme: default
        toc: yes
        toc_depth: 3
        toc_float:
            collapsed: no
            smooth_scroll: yes
---




```{r dependencies, include=FALSE}
library("ggplot2")
library("dplyr")
library("survival")
library("knitr")
library("gridExtra")
library("tidyr")
library("kableExtra")
library("survminer")
library("data.table")
options(warn=-1)
```

```{r read_files, include= F}
moms= read.table(params$maternal, h=T, sep= '\t')
dads= read.table(params$paternal, h=T, sep= '\t')
fets= read.table(params$fetal, h=T, sep= '\t')
```

```{r funk, include= F}

filter_df= function(df){
final= mutate(df, spont= as.numeric(FSTART==1 & (is.na(KSNITT) | KSNITT>1) &
                (is.na(KSNITT_PLANLAGT) | KSNITT_PLANLAGT==1) &
                INDUKSJON_PROSTAGLANDIN==0 & 
		INDUKSJON_ANNET==0 &
                INDUKSJON_OXYTOCIN==0 & 
		INDUKSJON_AMNIOTOMI==0), 
		PARITY0= as.numeric(PARITET_5==0))

return(final)
}

bin_funk= function(x){
bin= diff(range(x)) / (2 * IQR(x) / length(x)^(1/3))
return(bin)
}

```

# **Introduction**


# **Aims**
1. To describe ROH patterns in trios,
2. to investigate whether maternal, paternal and fetal ROH and
3. specific ROH areas of the genome are associated with spontaneous delivery.


# **Methods**

## Genotype

Analysis was performed in two Norwegian data sets (HARVEST and HUNT) linked to the Norwegian Medical Birth Register.

HARVEST samples (30000 trios) were genotyped using two different Illumina-Chip Array (HumanCoreExome12 and HumanCoreExome24), HUNT samples (25000 maternal, 15000 paternal and 15000 fetal samples) were genotyped with several chips.

Similar QC was applied to both datasets. Samples with major ethnicity other than CEU or cryptic relatedness to other samples were excluded (somwehere closer than second-third cousins).

## Phenotype

Only one pregnancy per sample was allowed. Only **singleton pregnancies**, with a **live birth** and a gestational age between 150 and 308 days were included. Artifitially conceived pregnancies, abruptio placentae, placenta previa, polyhydramnion and fetal malformations were excluded.

## ROH Calls

All analysis were carried out in separate for each data set and sample (maternal, paternal and fetal) for autosomal chromosomes. PLINK was used for ROH calling.  
Pruned SNPs list (R^2 > 0.9) to remove redundant markers in SNP-dense regions, and MAF filter > 0.01.

A ROH was defined according to [this](https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-12-460):  
        - 65 or more consecutive homozygous SNPs  
        - no heterozygote calls allowed within ROH  
        - ROH density > 1 SNP / 200 kb  
        - ROH broken in two if gap > 500 kb exists between adjacent homozygous SNPs  


Main exposure is total length of ROH (sum of bp for all ROH within individual)- scaled and centered.  


We also categorized ROHs according to their length:  
        - Short ROH <= 8.5 Mb  
        - Long ROH > 8.5 Mb (6th generation common ancestor)  

## Analysis

The effect of ROH on time-to-spontaneous delivery was estimated using Cox proportional hazards. Gestational age was defined as the time (either by UL - HARVEST or LMP - HUNT), and spontaneous delivery as event. C-sections and inductions were censored.



# **Runs of Homozygosity (ROH) description**

## ROH distribution by chromosome



### Distribution by chromosome

```{r echo=F}
m= fread('/mnt/work/pol/ROH/runs/harvest_m12_maternal.hom')
m24= fread('/mnt/work/pol/ROH/runs/harvest_m24_maternal.hom')

m= rbind(m, m24)
ms= m %>% group_by(CHR) %>% summarize(n= n(), KB= median(KB, na.rm=T))

d= fread('/mnt/work/pol/ROH/runs/harvest_m12_paternal.hom')
d24= fread('/mnt/work/pol/ROH/runs/harvest_m24_paternal.hom')

d= rbind(d, d24)
ds= d %>% group_by(CHR) %>% summarize(n= n(), KB= median(KB, na.rm=T))

fe= fread('/mnt/work/pol/ROH/runs/harvest_m12_fetal.hom')
fe24= fread('/mnt/work/pol/ROH/runs/harvest_m24_fetal.hom')

fe= rbind(fe, fe24)
fes= fe %>% group_by(CHR) %>% summarize(n= n(), KB= median(KB, na.rm=T))


x= ggplot(m, aes(factor(CHR), KB, fill= factor(CHR)))+ 
geom_violin() +
scale_fill_manual(values = rep(c('#a1dab4', '#41b6c4'), 11)) +
ylim(0, 5000) + 
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Chromosome') +
    ylab('Kb') +
ggtitle("Mothers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x1=  ggplot(d, aes(factor(CHR), KB, fill= factor(CHR)))+
geom_violin() +
scale_fill_manual(values = rep(c('#a1dab4', '#41b6c4'), 11)) +
ylim(0, 5000) +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Chromosome') +
    ylab('Kb') +
ggtitle("Fathers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))


x2= ggplot(fe, aes(factor(CHR), KB, fill= factor(CHR)))+
geom_violin() +
scale_fill_manual(values = rep(c('#a1dab4', '#41b6c4'), 11)) +
ylim(0, 5000) +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Chromosome') +
    ylab('Kb') +
ggtitle("Children") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))


grid.arrange(x, x1, x2)
```



### Length of ROH per chromosome

```{r echo=F}
x= ggplot(ms, aes(factor(CHR), KB, color= n)) +
geom_point() +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Chromosome') +
    ylab('Median (Kb)') +
ggtitle("Mothers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x1= ggplot(ds, aes(factor(CHR), KB, color= n)) +
geom_point() +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Chromosome') +
    ylab('Median (Kb)') +
ggtitle("Fathers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x2= ggplot(fes, aes(factor(CHR), KB, color= n )) +
geom_point() +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Chromosome') +
    ylab('Median (Kb)') +
ggtitle("Children") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

grid.arrange(x, x1, x2)

```



### Number of ROH per chromosome

```{r echo=F}

x= ggplot(ms, aes(factor(CHR), n, color= KB)) + 
geom_point() +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Chromosome') +
    ylab('N ROH') +
ggtitle("Mothers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x1= ggplot(ds, aes(factor(CHR), n, color= KB)) +
geom_point() +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Chromosome') +
    ylab('N ROH') +
ggtitle("Fathers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))


x2= ggplot(fes, aes(factor(CHR), n, color= KB)) +
geom_point() +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Chromosome') +
    ylab('N ROH') +
ggtitle("Children") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

grid.arrange(x, x1, x2)

```



## ROH means

```{r filter_data_sets, echo= F, message= F}

moms= filter_df(moms)
dads= filter_df(dads)
fets= filter_df(fets)

x= moms %>% summarize(mKB= mean(KB, na.rm=T), mNSEG= mean(NSEG, na.rm=T), nlongROH= mean(long_ROH_count, na.rm=T), mlongROH= mean(KB_long,na.rm=T), mshortROH= mean(KB_short, na.rm=T))
x1= dads %>% summarize(mKB= mean(KB, na.rm=T), mNSEG= mean(NSEG, na.rm=T), nlongROH= mean(long_ROH_count, na.rm=T), mlongROH= mean(KB_long,na.rm=T), mshortROH= mean(KB_short, na.rm=T))
x2= fets %>% summarize(mKB= mean(KB, na.rm=T), mNSEG= mean(NSEG, na.rm=T), nlongROH= mean(long_ROH_count, na.rm=T), mlongROH= mean(KB_long,na.rm=T), mshortROH= mean(KB_short, na.rm=T))



d= do.call("rbind", list(x, x2, x1))
d$role= c('Mother', 'Father', 'Child')


print(paste('Total number of mothers:', nrow(moms)))
print(paste('Total number of fathers:', nrow(dads)))
print(paste('Total number of children:', nrow(fets)))

d= d %>% select(role, mKB, mNSEG, nlongROH, mlongROH, mshortROH)

kable(d, col.names = c('Role', 'ROH length', 'N ROH', 'Prop. long ROH', 'Long ROH length', 'Short ROH length')) %>% kable_styling(bootstrap_options = c("striped", "hover"), full_width = F, position = "center", font_size = 12) %>%   column_spec(1, bold = T, border_right = T)

```



## ROH distribution

```{r echo=F, }

bins= bin_funk(moms$KB)
x= ggplot(moms, aes(KB)) +
geom_histogram(aes(y=..density..) , binwidth = bins, colour="#41b6c4", fill="#41b6c4") +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Total ROH length (Kb)') +
    ylab('Count') +
ggtitle("Mothers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

bins= bin_funk(dads$KB)

x1= ggplot(dads, aes(KB)) +
geom_histogram(aes(y=..density..) , binwidth = bins, colour="#41b6c4", fill="#41b6c4") +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Total ROH length (Kb)') +
    ylab('Count') +
ggtitle("Fathers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

bins= bin_funk(fets$KB)
x2= ggplot(fets, aes(KB)) +
geom_histogram(aes(y=..density..) , binwidth = bins,colour="#41b6c4", fill="#41b6c4") +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Total ROH length (Kb)') +
    ylab('Count') +
ggtitle("Children") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x

grid.arrange(x1, x2, nrow= 1)

```



## Correlations between number of ROH and total ROH length

```{r echo=F}
x= ggplot(moms, aes(NSEG, KB)) +
geom_point(alpha= 0.5, colour= '#41b6c4') +
theme_bw(base_family = "Source Sans Pro") +
theme(text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Number of ROH') +
    ylab('Total ROH length (Kb)') +
ggtitle("Mothers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x1= ggplot(dads, aes(NSEG, KB)) +
geom_point(alpha= 0.5, colour='#41b6c4') +
theme_bw(base_family = "Source Sans Pro") +
theme(text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Number of ROH') +
    ylab('Total ROH length (Kb)') +
ggtitle("Fathers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x2= ggplot(fets, aes(NSEG, KB)) +
geom_point(alpha= 0.5, colour='#41b6c4') +
theme_bw(base_family = "Source Sans Pro") +
theme(text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('Number of ROH') +
    ylab('Total ROH length (Kb)') +
ggtitle("Children") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x

grid.arrange(x1, x2, nrow= 1)

```

## ROH and PCA

```{r echo=F}

x= ggplot(moms, aes(PC1, KB, color= PC1)) +
geom_point() +
geom_smooth(method = "lm", se = FALSE, colour="#2c7fb8") +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('PC1') +
    ylab('Length of ROHs (Kb)') +
ggtitle("Mothers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x1= ggplot(dads, aes(PC1, KB, color= PC1)) +
geom_point() +
geom_smooth(method = "lm", se = FALSE, colour="#2c7fb8") +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('PC1') +
    ylab('Length of ROHs (Kb)') +
ggtitle("Fathers") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))


x2= ggplot(moms, aes(PC1, KB, color= PC1)) +
geom_point() +
geom_smooth(method = "lm", se = FALSE, colour="#2c7fb8") +
scale_colour_gradient(low="#a1dab4", high="#41b6c4") +
theme_bw(base_family = "Source Sans Pro") +
theme( text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()
    ) + xlab('PC1') +
    ylab('Length of ROHs (Kb)') +
ggtitle("Fets") +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

x

grid.arrange(x1, x2)

```


# **Main results**



## All ROH and spontaneous delivery


```{r echo=F}

fullcoxm= coxph(Surv(SVLEN_UL_DG, spont)~ scale(KB) + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0, moms)
fullcoxd= coxph(Surv(SVLEN_UL_DG, spont)~ scale(KB) + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0, dads)
fullcoxf= coxph(Surv(SVLEN_UL_DG, spont)~ scale(KB) + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0, fets)

x= summary(fullcoxm)$conf.int[1, c(1,3,4)]
x1= summary(fullcoxd)$conf.int[1, c(1,3,4)]
x2= summary(fullcoxf)$conf.int[1, c(1,3,4)]

d= do.call("rbind", list(x, x1, x2))
d= data.frame(d)
d$role= c('Mother', 'Father', 'Child')

d= d %>% select(role, exp.coef., lower..95, upper..95)

kable(d, col.names = c('Role', 'Hazard Ratio', 'Lower 95% CI', 'Upper 95% CI')) %>% kable_styling(bootstrap_options = c("striped", "hover"), full_width = F, position = "center", font_size = 12) %>%   column_spec(1, bold = T, border_right = T)

ggplot(d, aes(x= role, y= exp.coef.)) + 
geom_errorbar(aes(ymin= lower..95 , ymax= upper..95), width= 0.1) +
geom_point() +
theme_bw(base_family = "Source Sans Pro") +
theme(text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank()) + 
xlab('Sample') +
    ylab('HR for spontaneous delivery (95% CI)') +
ggtitle("ROH and Spontaneous delivery risk") +
geom_hline(aes(yintercept= 1), colour= 'grey', linetype= 'dashed') + 
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

```



## Maternal ROH and spontaneous delivery

```{r echo=F, fig.width= 12, fig.height= 10}

moms$QKB <- with(moms, cut(KB, breaks=quantile(KB, probs=seq(0,1, by=0.33), na.rm=TRUE), include.lowest=TRUE))

coxm= survfit(Surv(SVLEN_UL_DG, spont)~ QKB, moms)



p= ggsurvplot(
  coxm,                     # survfit object with calculated statistics.
  data = moms,  # data used to fit survival curves. 
  risk.table = TRUE,  
  palette = 
    c("gold", "#41b6c4","#253494"),# show risk table.
  pval = TRUE,             # show p-value of log-rank test.
  conf.int = FALSE,         # show confidence intervals for 
  # point estimaes of survival curves.
  xlim = c(250,300),        # present narrower X axis, but not affect
  # survival estimates.
  break.time.by = 10,     # break X axis in time intervals by 500.
  ggtheme = theme_minimal(base_family ='URWHelvetica'), # customize plot and risk table with a theme.
  risk.table.y.text.col = T, # colour risk table text annotations.
  risk.table.y.text = FALSE, # show bars instead of names in text annotations
  # in legend of risk table
  xlab = "Time in days",
  legend.labs = c("Low tertile", "Mid-tertile", "Higher Tertile"),  
  surv.median.line = "hv",
  censor=FALSE,
  font.family= 'URWHelvetica') 

p= ggpar(p,  font.title    = c(16, "bold", "#969696"),         
  font.subtitle = c(16, "bold.italic", "purple"), 
  font.x        = c(16, "plain", "black"),          
  font.y        = c(16, "plain", "black"),      
  font.xtickslab = c(14, "plain", "#969696"),
  #font.ytickslab = c(12, "plain", "#969696"),
  legend = "top"
)

print(p)
```



## Paternal ROH and spontaneous delivery


```{r echo=F, fig.width= 12, fig.height= 10}

dads$QKB <- with(dads, cut(KB, breaks=quantile(KB, probs=seq(0,1, by=0.33), na.rm=TRUE), include.lowest=TRUE))

coxd= survfit(Surv(SVLEN_UL_DG, spont)~ QKB, dads)



p= ggsurvplot(
  coxd,                     # survfit object with calculated statistics.
  data = dads,  # data used to fit survival curves. 
  risk.table = TRUE,
  palette =
    c("gold", "#41b6c4","#253494"),# show risk table.
  pval = TRUE,             # show p-value of log-rank test.
  conf.int = FALSE,         # show confidence intervals for 
  # point estimaes of survival curves.
  xlim = c(250,300),        # present narrower X axis, but not affect
  # survival estimates.
  break.time.by = 10,     # break X axis in time intervals by 500.
  ggtheme = theme_minimal(base_family ='URWHelvetica'), # customize plot and risk table with a theme.
  risk.table.y.text.col = T, # colour risk table text annotations.
  risk.table.y.text = FALSE, # show bars instead of names in text annotations
  # in legend of risk table
  xlab = "Time in days",
  legend.labs = c("Low tertile", "Mid-tertile", "Higher Tertile"),
  surv.median.line = "hv",
  censor=FALSE,
  font.family= 'URWHelvetica')

p= ggpar(p,  font.title    = c(16, "bold", "#969696"),
  font.subtitle = c(16, "bold.italic", "purple"),
  font.x        = c(16, "plain", "black"),
  font.y        = c(16, "plain", "black"),
  font.xtickslab = c(14, "plain", "#969696"),
  #font.ytickslab = c(12, "plain", "#969696"),
  legend = "top"
)

print(p)
```



## Children ROH and spontaneous delivery

```{r echo=F, fig.width= 12, fig.height= 10}

fets$QKB <- with(fets, cut(KB, breaks=quantile(KB, probs=seq(0,1, by=0.33), na.rm=TRUE), include.lowest=TRUE))

coxf= survfit(Surv(SVLEN_UL_DG, spont)~ QKB, fets)



p= ggsurvplot(
  coxf,                     # survfit object with calculated statistics.
  data = fets,  # data used to fit survival curves. 
  risk.table = TRUE,
  palette =
    c("gold", "#41b6c4","#253494"),# show risk table.
  pval = TRUE,             # show p-value of log-rank test.
  conf.int = FALSE,         # show confidence intervals for 
  # point estimaes of survival curves.
  xlim = c(250,300),        # present narrower X axis, but not affect
  # survival estimates.
  break.time.by = 10,     # break X axis in time intervals by 500.
  ggtheme = theme_minimal(base_family ='URWHelvetica'), # customize plot and risk table with a theme.
  risk.table.y.text.col = T, # colour risk table text annotations.
  risk.table.y.text = FALSE, # show bars instead of names in text annotations
  # in legend of risk table
  xlab = "Time in days",
  legend.labs = c("Low tertile", "Mid-tertile", "Higher Tertile"),
  surv.median.line = "hv",
  censor=FALSE,
  font.family= 'URWHelvetica')

p= ggpar(p,  font.title    = c(16, "bold", "#969696"),
  font.subtitle = c(16, "bold.italic", "purple"),
  font.x        = c(16, "plain", "black"),
  font.y        = c(16, "plain", "black"),
  font.xtickslab = c(14, "plain", "#969696"),
  #font.ytickslab = c(12, "plain", "#969696"),
  legend = "top"
)

print(p)
```



# **Secondary analysis**

## Long and short ROH

```{r echo=F}
fullcoxml= coxph(Surv(SVLEN_UL_DG, spont)~ scale(KB_long) + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0, moms)
fullcoxdl= coxph(Surv(SVLEN_UL_DG, spont)~ scale(KB_long) + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0, dads)
fullcoxfl= coxph(Surv(SVLEN_UL_DG, spont)~ scale(KB_long) + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0, fets)

x= summary(fullcoxml)$conf.int[1, c(1,3,4)]
x1= summary(fullcoxdl)$conf.int[1, c(1,3,4)]
x2= summary(fullcoxfl)$conf.int[1, c(1,3,4)]

d= do.call("rbind", list(x, x1, x2))
d= data.frame(d)
d$role= c('Mother', 'Father', 'Child')
d$bp= 'Long ROH'
d= d %>% select(role, bp, exp.coef., lower..95, upper..95)

fullcoxms= coxph(Surv(SVLEN_UL_DG, spont)~ scale(KB_short) + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0, moms)
fullcoxds= coxph(Surv(SVLEN_UL_DG, spont)~ scale(KB_short) + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0, dads)
fullcoxfs= coxph(Surv(SVLEN_UL_DG, spont)~ scale(KB_short) + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + PARITY0, fets)

x= summary(fullcoxms)$conf.int[1, c(1,3,4)]
x1= summary(fullcoxds)$conf.int[1, c(1,3,4)]
x2= summary(fullcoxfs)$conf.int[1, c(1,3,4)]

ds= do.call("rbind", list(x, x1, x2))
ds= data.frame(ds)
ds$role= c('Mother', 'Father', 'Child')
ds$bp= 'Short ROH'
ds= ds %>% select(role, bp, exp.coef., lower..95, upper..95)

d= rbind(d, ds)


#kable(d, col.names = c('Role', 'Hazard Ratio', 'Lower 95% CI', 'Upper 95% CI')) %>% kable_styling(bootstrap_options = c("striped", "hover"), full_width = F, position = "center", font_size = 12) %>%   column_spec(1, bold = T, border_right = T)

ggplot(d, aes(x= bp, y= exp.coef.)) +
facet_grid(. ~ role) +
geom_errorbar(aes(ymin= lower..95 , ymax= upper..95), width= 0.1) +
geom_point() +
theme_bw(base_family = "Source Sans Pro") +
theme(text = element_text(size= 14),
           legend.position="none",
           panel.border = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank(),
	strip.background =element_rect(fill="white")) + 
xlab('Sample') +
    ylab('HR for spontaneous delivery (95% CI)') +
ggtitle("ROH and Spontaneous delivery risk") +
geom_hline(aes(yintercept= 1), colour= 'grey', linetype= 'dashed') +
    theme(axis.line.x = element_line(color="black", size = 0.2),
          axis.line.y = element_line(color="black", size = 0.2))

```


# **Interpretation**


The HR for maternal ROH length was 1.03 [95% CI: 1.01,1.06], which indicates that at any time, mothers with larger total ROH had a greater risk of delivering spontaneously. We found no effect of fetal nor paternal ROH lengths. 

Recent inbreeding (<6 generations) results were more similar to those found with total ROH length (HR: 1.04; 95%: 1.01, 1.06).  



# **To be done...**

1. Confounding (sociodemographic factors?)
2. Replication in HUNT  
3. Mapping of homozygous SNPs (GWAS on SNPs conditioned to pertenance to ROH)  
4. Shared maternal-fetal ROHs?
5. What else?

```{r echo=F}
sessionInfo()
```
